### DISCUSSION TITLE
Power of 2 & Jolt Support

### DISCUSSION THREAD

**Discussion by TokisanGames** [TokisanGames/Terrain3D#152] [Architecture]
I tested Jolt yesterday. It requires power of 2 sized heightmap collision shapes while ours are power of 2 + 1. After making small changes, it ran fine, but upon close inspection I noticed a few issues. If I can get help resolving these two issues I'm happy to go down to 1024^2 heightmap collision shapes.

Godot-Jolt Issue: Jolt supporting non-power of 2 sized heightmap collisionshapes.
https://github.com/godot-jolt/godot-jolt/issues/502

## 1. Gaps Between Collision Shapes

We have 1024^2 sized heightmaps. Our [HeightmapShape3D collision shape](https://docs.godotengine.org/en/stable/classes/class_heightmapshape3d.html) has a size of 1025^2. 

The reason for this is each pixel on the heightmap corresponds to one vertex on the collision shape. The last row (1025th) is there to overlap the gap between heightmaps. It should be set to the same value as the first row on the next image. You can see on this image, if we had only 1024 vertices, we would have only 1023 quads and would have a gap between 1023 and 0 on the next. The last row is in purple.

![image](https://github.com/outobugi/Terrain3D/assets/632766/ab47d2ff-2cf6-4626-ab81-f1a65f234cb4)

You can see this in practice. Collision is white. When we reduce our collision shapes down to 1024 we are left with gaps between regions. I'm not sure how to fix this without going to 1025 or ignoring the last row of the heightmap, wasting vram. I'm open to solutions and help.

![image](https://github.com/outobugi/Terrain3D/assets/632766/e58cc5ce-1617-4e88-ae43-b4581aa531d3)

## 2. Mesh/Collision Alignment

We have 1024^2 heightmaps, 1025^2 collision shape. The shader uses a `vec2(-0.5, -0.5)` offset on the texel lookup (still not sure why its required), and we can achieve pixel perfect alignment between the mesh and the collision shape. Subtle white lines are collision.

![image](https://github.com/outobugi/Terrain3D/assets/632766/2cf3e0af-4662-4eab-9a71-c9b1cd149a63)

![image](https://github.com/outobugi/Terrain3D/assets/632766/c871cbdc-8038-4613-a665-a4c6dd445ebd)

I reduced the collisionshape down to 1024, and ignoring the gaps for now, I looked at alignment. I found if I removed the texel offset, the collision mostly lines up with the mesh, however it is not pixel perfect. It is worse with the offset.

1025^2 on the left. 1024^2 on the right

![image](https://github.com/outobugi/Terrain3D/assets/632766/6f4483bc-cb0b-4bb5-99fc-ec20b4893252)

Heightmap generation. The only change made is removing the +1 from L121 `shape_size = region_size;`.
https://github.com/outobugi/Terrain3D/blob/41f5456e4f5eb923738b55c66d275194d7e35408/src/terrain_3d.cpp#L120-L140

Vertex shader:
The only change is L126 removing the first `-vec2(0.5);`
https://github.com/outobugi/Terrain3D/blob/41f5456e4f5eb923738b55c66d275194d7e35408/src/shaders/main.glsl#L123-L129

I'm not sure how to get it to align properly and would appreciate any help.

**Comment by jrouwe** [comment]
Hello,

What we have been doing is to see the height field samples as the value at a vertex location like so:

![image](https://github.com/outobugi/Terrain3D/assets/1621693/3ceab3a8-f90b-4394-858d-bd88cebde051)

This shows 2 connecting 4x4 heightfields, 3x3 quads. The pixels in the rows where they are touching must have the same value in order for the height fields to connect without seams.

From your first image it looks like you see the height samples as the height at the center of a quad rather than at a vertex. Does this mean that if you're rendering at the edge of a tile that you need to sample from 2 or more textures in the vertex shader to create a continuous mesh? It would be helpful if you draw how the data is being interpreted.

**Reply by TokisanGames** [reply]
Hi, thanks for the discussion. 

I think your image shows exactly what we are doing. Height samples of the texture correspond to physics shape vertices, not the center of quads. 

On your image, you have a 4x4 physics shape and a 3x3 heightmap texture. Here I've drawn several 3x3 heightmap textures. The last row/col on the heighmap shape must read from the next adjacent texture so the height values match. You can see these two 4x4 collision shapes require values from 6 different textures. The position of the physics shape in the colored boxes is unimportant, they just represent pixels. 1 colored pixel = 1 heightmap shape vertex/dot.
![image](https://github.com/outobugi/Terrain3D/assets/632766/400c01c8-4f10-4285-893a-25324b8d7913)

However, our textures are powers of 2 rather than the physics shapes. If we put your drawing with our numbers, then instead of 4 blue dots, it's 1025 blue dots. Instead of 3 quads, it's 1024 quads. Thus we have a 1024x1024 texture, and a 1025x1025 collision. If our collision goes down to 1024, then we'll have a gap. On your image, if you reduce the blue dots physics shape down to 3x3, you'll be missing a row of quads between the edge of the last blue dot and the start of the first orange dot. 

That's issue #1 I highlighted above. It seems we're interpreting it the same way, 1 image pixel = 1 physics shape vertex. That means the physics shape needs one more to bridge the gap to the next shape. So then the question is which is more optimal for a game/engine as a whole, optimizing the physics engine on the CPU for powers of 2 and leaving textures to be N^2-1, or optimizing the system for the GPU with N^2 textures?

**Reply by jrouwe** [reply]
Would it not be beneficial for GPU when using N^2-1 textures to use that last pixel to replicate the value of the next tile? That way you never need to read from neighboring textures and to render these 2 tiles you only need 2 textures instead of 6.

**Reply by TokisanGames** [reply]
We're talking about setting the heightmap shape which is processing on the CPU, so no that won't help the GPU. 

In fact it will waste VRAM by not using that strip on every texture. 

We have a 16 x 16 1024^2 terrain, so that's 256 texture layers, *3 for height, color and control maps. Later we'll have detail maps. If you want each one to be 1023^2 then we're wasting 500k pixels across all maps, about 6MB of vram. I don't know about the performance of texture lookups across odd sized textures. 

Also we're talking about expanding it much larger, up to 2048 layers, so much more waste. E.g. up to 45 x 45 2048^2 or maybe 4096^2. 

As it is, power of 2 textures is ideal for the GPU, and our CPU processing is straight forward. One pixel per vertex. Little edge case code. 

We'll have to write a lot of edge case code to manage the uneven sizes both in c++ and the shader. Each region can be edited, added, or removed independently, so we'll be editing multiple texture maps simultaneously to duplicate pixels sometimes, not at other times. It will be an unwieldy task to meet your suggestion. 

So unwieldy that I don't see it being worth the effort of redesigning our system for no clear benefit. If I only had power of two heightmap shapes, I would either use them, then create separate collision shapes to fill in the gaps between them. Or I wouldn't use them at all and instead generate our own collision shapes that are sized appropriately (1025).

**Reply by jrouwe** [reply]
Ok, I will investigate height fields of size 2^N+1 and see if it is possible to support this.

**Reply by TokisanGames** [reply]
Great, thanks. I look forward to seeing what you find.

**Reply by Dabnabbit** [reply]
I'm also following this exact same issue attempting to implement Jolt physics. Same exact reasonings essentially, look forward to hearing any updates!

**Reply by jrouwe** [reply]
I just submitted a change that allows non-power of 2 height fields. Note: godot-jolt still needs to be updated with this new version of Jolt.

**Reply by mihe** [reply]
With this new change it seems like Terrain3D's demo project works just fine. Expect this to be included in the next release of Godot Jolt, which should hopefully be out some time early next week.

**Reply by TokisanGames** [reply]
Excellent. Thanks guys. This will help out several projects.

**Reply by jrouwe** [reply]
I think this issue can be closed now, see: https://github.com/godot-jolt/godot-jolt/issues/502#issuecomment-1677028815

**Reply by TokisanGames** [reply]
Thank you
